hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java

```diff
@@ -168,6 +225,43 @@ public int getTrackedNodeCount() {
     return trackedNodes.size();
   }
 
+  synchronized void setMetricsToGauge() {
+    metrics.setTrackedContainersUnhealthyTotal(unhealthyContainers);
+    metrics.setTrackedRecommissionNodesTotal(trackedRecommission);
+    metrics.setTrackedDecommissioningMaintenanceNodesTotal(
+            trackedDecomMaintenance);
+    metrics.setTrackedContainersUnderReplicatedTotal(
+            underReplicatedContainers);
+    metrics.setTrackedContainersSufficientlyReplicatedTotal(
+            sufficientlyReplicatedContainers);
+    metrics.setTrackedPipelinesWaitingToCloseTotal(pipelinesWaitingToClose);
+    for (Map.Entry<String, Long> e :
+            pipelinesWaitingToCloseByHost.entrySet()) {
+      metrics.metricRecordPipelineWaitingToCloseByHost(e.getKey(),
+              e.getValue());
+    }
+    for (Map.Entry<String, ContainerStateInWorkflow> e :
```

Comment:
I might be wrong, but I think there is a bug here.

Lets say we put a host to maintenance. It will have some metrics tracked in the ByHost maps.

After each pass we reset these maps to have zero counts, but we don't remove the entries from the maps anywhere (unless I have missed it). Then we update the values accordingly.

Later the node goes back into service and even though it is removed from the monitor, it will be tracked with zero counts forever.

Over time on a long running cluster, we will build up a lot of "by host" metrics with zero values, when they really should be removed.

I think the reset will need to remove them from the maps rather than zeroing them, and also when setting the values to the metric gauge, you will need to remove values no longer there from it too.

It might be easier to pass a `Map<String, ContainerStateInWorkflow>` to the metrics class to facilitate removing the stale entries.

Reply:
Huum, looks like you are correct. I wonder what the best approach is here.

I don't think its a great user experience if we start with no individual nodes track, and then over time (in a long running SCM) more and more nodes get added for maintenance and decommission and the number builds up all with zero counts. I guess its not a major problem, but it would be nice to resolve it somehow.

Reply:
In #3791 Symious added a tag with a group of metrics in JSON form. For the metrics system, this is just a tag to string, rather than a gauge, but we could group all currently decommissioning / maintenence nodes into a JSON representation to expose the fine grained info. If no nodes are in the workflow, it would just be an empty json object, so nodes can come and go easily.

Then you still have your aggregate metrics as they are now.

It is unlikely that someone would want to chart an individual DN as they would have to create a new chart for each DN.

What do you think?

Reply:
I've modified the code to dynamically (without using the helper `MetricsRegistry` class to add gauges) add to the collector as is done similarly in namenode topmetrics collections.  See https://github.com/apache/hadoop/blob/eefa664fea1119a9c6e3ae2d2ad3069019fbd4ef/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/top/metrics/TopMetrics.java#L167.
Here the metrics are collected dynamically by host when the host node is in the workflow.  When the node exits the workflow, the metrics for that host are no longer collected.  In the `JMX,` the node metrics are no longer in the output.  Note this is true for `JMX`, the prom endpoint _seems to retain_ the last value pushed.  See the following metrics pushed out to JMX for the` NodeDecommissionMetrics` when `datanode-2` is decommissioned:

```
before

"name" : "Hadoop:service=StorageContainerManager,name=NodeDecommissionMetrics",
    "modelerType" : "NodeDecommissionMetrics",
    "tag.Hostname" : "0d207b6cbbf1",
    "TrackedDecommissioningMaintenanceNodesTotal" : 0,
    "TrackedRecommissionNodesTotal" : 0,
    "TrackedPipelinesWaitingToCloseTotal" : 0,
    "TrackedContainersUnderReplicatedTotal" : 0,
    "TrackedContainersUnhealthyTotal" : 0,
    "TrackedContainersSufficientlyReplicatedTotal" : 0
  }, {

during
    "name" : "Hadoop:service=StorageContainerManager,name=NodeDecommissionMetrics",
    "modelerType" : "NodeDecommissionMetrics",
    "tag.Hostname" : "0d207b6cbbf1",
    "TrackedDecommissioningMaintenanceNodesTotal" : 1,
    "TrackedRecommissionNodesTotal" : 0,
    "TrackedPipelinesWaitingToCloseTotal" : 2,
    "TrackedContainersUnderReplicatedTotal" : 0,
    "TrackedContainersUnhealthyTotal" : 0,
    "TrackedContainersSufficientlyReplicatedTotal" : 0,
    "TrackedUnhealthyContainers-ozone-datanode-2.ozone_default" : 0,
    "TrackedSufficientlyReplicated-ozone-datanode-2.ozone_default" : 0,
    "TrackedPipelinesWaitingToClose-ozone-datanode-2.ozone_default" : 2,
    "TrackedUnderReplicated-ozone-datanode-2.ozone_default" : 0
  }, {

  }, {
    "name" : "Hadoop:service=StorageContainerManager,name=NodeDecommissionMetrics",
    "modelerType" : "NodeDecommissionMetrics",
    "tag.Hostname" : "0d207b6cbbf1",
    "TrackedDecommissioningMaintenanceNodesTotal" : 1,
    "TrackedRecommissionNodesTotal" : 0,
    "TrackedPipelinesWaitingToCloseTotal" : 0,
    "TrackedContainersUnderReplicatedTotal" : 1,
    "TrackedContainersUnhealthyTotal" : 0,
    "TrackedContainersSufficientlyReplicatedTotal" : 0,
    "TrackedUnhealthyContainers-ozone-datanode-2.ozone_default" : 0,
    "TrackedSufficientlyReplicated-ozone-datanode-2.ozone_default" : 0,
    "TrackedPipelinesWaitingToClose-ozone-datanode-2.ozone_default" : 0,
    "TrackedUnderReplicated-ozone-datanode-2.ozone_default" : 1
  }, {

after
 }, {
    "name" : "Hadoop:service=StorageContainerManager,name=NodeDecommissionMetrics",
    "modelerType" : "NodeDecommissionMetrics",
    "tag.Hostname" : "0d207b6cbbf1",
    "TrackedDecommissioningMaintenanceNodesTotal" : 0,
    "TrackedRecommissionNodesTotal" : 0,
    "TrackedPipelinesWaitingToCloseTotal" : 0,
    "TrackedContainersUnderReplicatedTotal" : 0,
    "TrackedContainersUnhealthyTotal" : 0,
    "TrackedContainersSufficientlyReplicatedTotal" : 0
  }, {

```
The host `datanode-2 `metrics no longer visible as the node exits the workflow.

This seems to follow how hadoop handles metrics collected dynamically, however the `prom endpoint` seems to _**retain**_ the last pushed value for some reason.  Is this what we should expect when collecting metrics for hosts as they go in and out of the workflow?

Reply:
I'm not sure how the prom end point works. Its not ideal that it keeps the last value pushed, but I am not sure where that code even comes from!

Reply:
Thanks.  We should go forward with using this implementation that works for JMX metrics for completing this PR  to _expose decommission / maintenance metrics via JMX_ and open a new jira to look into supporting the prom endpoint.  This PR supports metrics tracking the decommission and maintenance workflow both with aggregated counts and DN host specific counts.  A jira will be filed to track prom endpoint behavior for the metrics.  What do you think?